### PR TITLE
feat: ESLint rule - all ngrx "Fail" actions should have `implements ErrorAction`

### PR DESCRIPTION
**Pull request by Platonn**
Goal: ensure that each NgRx Failure Action in our repo has an `error` property. It's ensured by enforcing each such action to have `implements ErrorAction`.

Means: 
- added ESLint rule which marks a class as invalid when it has `/Fail/` in its name but is missing `implements ErrorAction`
  - it's coped ONLY to files with a suffix `*.action*.ts` ( supposed to be files with NgRx actions )
    
Moreover, the ESLint rule provides an auto-fixer which:
- adds missing `implements ErrorAction`
- imports `ErrorAction` from `@sparatcus/core` in the current file

Note: to run unit tests of the ESLint rule, execute:
```bash
cd tools/eslint-rules
npx nx run test --watch
```

fixes https://jira.tools.sap/browse/CXSPA-7756

### CODE REVIEW

File: tools/eslint-rules/rules/no-ngrx-fail-action-without-error-action-implementation.spec.ts [hunk 1]
```diff
@@ -10,8 +10,28 @@ const ruleTester = new TSESLint.RuleTester({
...
+export class LoadProductFail implements Action, SomeOtherInterface, ErrorAction {}
+    `,
+    `
+    import { ErrorAction } from '@spartacus/core';
+export class LoadProductFail extends SuperClass implements ErrorAction {}
+    `,
+    `
+    import { ErrorAction } from '@spartacus/core';
+export class LoadProductFail extends SuperClass implements SomeOtherInterface, ErrorAction {}
+    `,
```
Comment (Platonn) [inline tools/eslint-rules/rules/no-ngrx-fail-action-without-error-action-implementation.spec.ts]: Let's add unit tests proving that _it doesnt matter if ErrorAction interface is implemented_ for actions that _don't have Fail in the name_.

```ts
    // actions with `Fail` in name, that implement `ErrorAction`
    ///// <ALREADY IMPLEMENTED VALID CASES HERE> /////
    
    // actions without `Fail` in name, that implement `ErrorAction`
    `        
    import { ErrorAction } from '@spartacus/core';
    export class LoadProduct implements ErrorAction {}
    `,
    `
    import { ErrorAction } from '@spartacus/core';
    export class LoadProduct implements Action, ErrorAction {}
    `,
    `
    import { ErrorAction } from '@spartacus/core';
    export class LoadProduct implements Action, SomeOtherInterface, ErrorAction {}
    `,
    `
    import { ErrorAction } from '@spartacus/core';
    export class LoadProduct extends SuperClass implements ErrorAction {}
    `,
    `
    import { ErrorAction } from '@spartacus/core';
    export class LoadProduct extends SuperClass implements SomeOtherInterface, ErrorAction {}
    `,
    
    // actions without `Fail` in name, that don't implement `ErrorAction`
    `        
    export class LoadProduct {}
    `,
    `
    export class LoadProduct implements Action {}
    `,
    `
    export class LoadProduct implements Action, SomeOtherInterface {}
    `,
    `
    export class LoadProduct extends SuperClass {}
    `,
    `
    export class LoadProduct extends SuperClass implements SomeOtherInterface {}
    `,
```

File: tools/eslint-rules/rules/utils/import-utils.spec.ts [hunk 2]
```diff
@@ -1 +1,98 @@
...
+  const ruleTester = new TSESLint.RuleTester({
+    parser: require.resolve('@typescript-eslint/parser'),
+    parserOptions: {
+      ecmaVersion: 2020,
+      sourceType: 'module',
+    },
+  });
+
+  // Define tests
+  ruleTester.run('has-implements-interface', rule, {
```
Comment (Platonn) [inline tools/eslint-rules/rules/utils/import-utils.spec.ts]: [minor]
leftover, what about `isIdentifierImported`  - like the name of the main function?

PS. we might want to also change `has-implements-interface` description in the other file to the function name `hasImplementsInterface`.

File: feature-libs/cart/base/core/store/actions/cart-voucher.action.ts [hunk 3]
```diff
@@ -42,6 +45,7 @@ export class CartAddVoucherFail extends StateUtils.EntityFailAction {
     }
   ) {
     super(PROCESS_FEATURE, ADD_VOUCHER_PROCESS_ID, payload.error);
+    this.error = payload.error;
```
Comment (Platonn) [inline feature-libs/cart/base/core/store/actions/cart-voucher.action.ts]: [question]
Do we really need this line? I can see we don't need it e.g. in `CreateWishListFail`. But both those actions `extends StateUtils.EntityFailAction` - which does the job of assigning value to the public `error` property.

Note: analogical question applies to other places in this PR

File: feature-libs/product-configurator/rulebased/core/state/actions/configurator.action.ts [hunk 4]
```diff
@@ -185,6 +185,7 @@ export class UpdateConfigurationFinalizeSuccess extends StateUtils.EntitySuccess
   }
 }
 
+// eslint-disable-next-line @nx/workspace/no-ngrx-fail-action-without-error-action-implementation
```
Comment (Platonn) [inline feature-libs/product-configurator/rulebased/core/state/actions/configurator.action.ts]: [minor]
Let's add explanation why it's ignored here after `... -- ...` + leave JIRA ticket number here, for reference, to sort it out in the near future.

And let's contact Team Tiger to take a look at it

File: projects/core/src/state/utils/entity-loader/entity-loader.action.ts [hunk 5]
```diff
@@ -102,12 +104,9 @@ export class EntityFailAction implements EntityLoaderAction {
...
    */
   constructor(entityType: string, id: EntityId);
-  constructor(
-    entityType: string,
-    id: EntityId,
-    public error?: any
-  ) {
+  constructor(entityType: string, id: EntityId, error?: any) {
     this.meta = entityFailMeta(entityType, id, error);
+    this.error = error;
```
Comment (Platonn) [inline projects/core/src/state/utils/entity-loader/entity-loader.action.ts:109]: [note]
Here it's OK, to implement this line. Thanks to it, all actions extending this class won't need to implement analogical mapping for `this.error` property, I guess.

Same for: `LoaderFailAction` and `EntityScopedFailAction`

File: tools/eslint-rules/rules/no-ngrx-fail-action-without-error-action-implementation.spec.ts [hunk 6]
```diff
@@ -10,8 +10,28 @@ const ruleTester = new TSESLint.RuleTester({
 });
 
 ruleTester.run(RULE_NAME, rule, {
-  // SPIKE TODO: ADD VALID ONES
-  valid: [],
+  valid: [
+    `        
+    import { ErrorAction } from '@spartacus/core';
+export class LoadProductFail implements ErrorAction {}
```
Comment (Platonn) [inline tools/eslint-rules/rules/no-ngrx-fail-action-without-error-action-implementation.spec.ts]: [nitpick]
For `valid` test cases, I believe we don't need this _broken indentation_. I hope we can align it with the `import` statement above, in terms of tabs/spaces

File: feature-libs/cart/base/core/store/actions/cart-voucher.action.ts (hunk 3 above)
Comment (pawelfras) [reply feature-libs/cart/base/core/store/actions/cart-voucher.action.ts]: It's a leftover. I made this change before modifying `EntityFailAction`.

File: feature-libs/product-configurator/rulebased/core/state/actions/configurator.action.ts (hunk 4 above)
Comment (pawelfras) [reply feature-libs/product-configurator/rulebased/core/state/actions/configurator.action.ts]: Sure, here's the follow-up ticket: https://jira.tools.sap/browse/CXSPA-7901

File: feature-libs/product-configurator/rulebased/core/state/actions/configurator.action.ts [hunk 7]
```diff
@@ -185,6 +185,12 @@ export class UpdateConfigurationFinalizeSuccess extends StateUtils.EntitySuccess
   }
 }
 
+/*
+ * UpdateConfigurationFinalFail contains `Fail` word in its name, but the purpose ot this action is not clear
+ * due to extending `EntitySuccessAction` instead of `EntitySuccessAction`
+ * That said, it's not know whether this action should implement `ErrorAction` or not.
+ * Follow-up ticket: CXSPA-7091
+ */
```
Comment (Platonn) [inline feature-libs/product-configurator/rulebased/core/state/actions/configurator.action.ts]: 1. Are you sure we want it to be an official JSDoc documentation of the class? (by using JSDoc - it looks like an intended, perhaps wasn't intended). Or it's just a developer's note?

2. explanations of eslint-disable, can be passed after `--` in the same line where the eslint-disable comment is

File: feature-libs/cart/base/core/store/actions/cart.action.ts [hunk 8]
```diff
@@ -58,10 +58,15 @@ interface CreateCartFailPayload extends CreateCartPayload {
...
-export class CreateCartFail extends StateUtils.EntityFailAction {
+export class CreateCartFail
+  extends StateUtils.EntityFailAction
+  implements ErrorAction
+{
   readonly type = CREATE_CART_FAIL;
+  error: any;
   constructor(public payload: CreateCartFailPayload) {
     super(MULTI_CART_DATA, payload.tempCartId, payload.error);
+    this.error = payload.error;
```
Comment (Platonn) [inline feature-libs/cart/base/core/store/actions/cart.action.ts]: Besides fixing the [comment](https://github.com/SAP/spartacus/pull/19042#discussion_r1680691373) , above I can still see many occurences of :
- declaring property `error: any` explicitly in the child class
- `this.error = <something from constructor argument>` in the child class
... while it doesn't seem to be needed, thanks to the `extends SomeSuperClass` which does the job

File: feature-libs/product-configurator/rulebased/core/state/actions/configurator.action.ts (hunk 7 above)
Comment (pawelfras) [reply feature-libs/product-configurator/rulebased/core/state/actions/configurator.action.ts]: 1) It's a multi-line comment and it won't appear as JSDoc. An example: 
image

2) Thanks, I wasn't aware. I can do it the way you proposed.

File: feature-libs/cart/base/core/store/actions/cart.action.ts (hunk 8 above)
Comment (pawelfras) [reply feature-libs/cart/base/core/store/actions/cart.action.ts]: Apologies for overlooking these places. Done!

**Review by Platonn — COMMENTED** [review]
✅ LGTM

**Review by pawelfras — APPROVED** [review]
LGTM too!

**Pull request merged**